client/rest/src/main/java/org/elasticsearch/client/NodeSelector.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Selects nodes that can receive requests. Use with
+ * {@link RestClientActions#withNodeSelector withNodeSelector}.
+ */
+public interface NodeSelector {
+    /**
+     * Select the {@link Node}s to which to send requests.
+     *
+     * @param nodes an unmodifiable list of {@linkplain Node}s in the order
+     *      that the {@link RestClient} would prefer to use them
+     * @return a subset of the provided list of {@linkplain Node}s that the
+     *      selector approves of, in the order that the selector would prefer
+     *      to use them.
+     */
+    List<Node> select(List<Node> nodes);
+
+    /**
+     * Selector that matches any node.
+     */
+    NodeSelector ANY = new NodeSelector() {
+        @Override
+        public List<Node> select(List<Node> nodes) {
+            return nodes;
+        }
+
+        @Override
+        public String toString() {
+            return "ANY";
+        }
+    };
+
+    /**
+     * Selector that matches any node that has metadata and doesn't
+     * have the {@code master} role OR it has the data {@code data}
+     * role. It does not reorder the nodes sent to it.
+     */
+    NodeSelector NOT_MASTER_ONLY = new NodeSelector() {
+        @Override
+        public List<Node> select(List<Node> nodes) {
+            List<Node> subset = new ArrayList<>(nodes.size());
+            for (Node node : nodes) {
+                if (node.getRoles() == null) continue;
+                if (false == node.getRoles().hasMasterEligible() || node.getRoles().hasData()) {
+                    subset.add(node);
+                }
+            }
+            return subset;
```

Comment:
Maybe we should try and make this a bit smarter? In the extreme case that there are no alive non-master only nodes around, do we want to make the client resurrect one of the dead nodes (I'd say we should rather throw exception when we return empty from here)? Maybe we should rather use a master-only node in this specific case? While I write I am actually not sure about this. Maybe instead of overloading the master eligible nodes we should just bail :) But document this behaviour clearly?

Reply:
I worry that "smarter" become "hard to understand" in this case. I think it makes sense to have the built in rule be hard and fast and if folks want to be more flexible in their application they can be.

One issue is that it is kind of hard to tell if you are being called with a list of live nodes or blacklisted ones. We could pass a boolean to communicate that, but that feels pretty complicated.

Reply:
you are right, this is good enough, and I agree at this point that we should try and revive a node if the selector returns empty.